gnssanalysis/gn_diffaux.py

```diff
@@ -669,3 +671,155 @@ def rac_df_to_rms_df(rac_df):
+) -> _pd.DataFrame:
+    """
+    Compare two SP3 files to calculate orbit and clock differences. The orbit differences will be represented
+    in both X/Y/Z ECEF frame and R/A/C orbit frame, and the clock differences will NOT be normalised.
+
+    :param _Path base_sp3_file: Path of the baseline SP3 file
+    :param _Path test_sp3_file: Path of the test SP3 file
+    :return _pd.DataFrame: The Pandas DataFrame containing orbit and clock differences
+    """
+    base_sp3_df = _gn_io.sp3.read_sp3(str(base_sp3_file))
+    test_sp3_df = _gn_io.sp3.read_sp3(str(test_sp3_file))
+
+    # Select rows with matching indices and calculate XYZ differences (ECEF)
+    common_indices = base_sp3_df.index.intersection(test_sp3_df.index)
+    diff_est_df = (
+        test_sp3_df.loc[common_indices, "EST"] - base_sp3_df.loc[common_indices, "EST"]
+    )
+
+    # Extract clocks and change the units from ms to ns (read_sp3 will result in sp3 units (ms))
+    # TODO: normalise clocks
+    diff_clk_df = diff_est_df["CLK"].to_frame(name="CLK") * 1e3
+
+    # Drop clocks and then change the units from km to m (read_sp3 will result in sp3 units (km))
+    diff_xyz_df = diff_est_df.drop(columns=["CLK"]) * 1e3
+
+    # RAC difference
+    # TODO: hlm_mode
+    diff_rac_df = _gn_io.sp3.diff_sp3_rac(base_sp3_df, test_sp3_df, hlm_mode=None)
+
+    # Drop the not-particularly needed 'EST_RAC' multi-index level
+    diff_rac_df.columns = diff_rac_df.columns.droplevel(0)
+
+    # Change the units from km to m (diff_sp3_rac will result in sp3 units (km))
+    diff_rac_df = diff_rac_df * 1e3
+
+    diff_sp3_df = diff_xyz_df.join(diff_rac_df)
+    diff_sp3_df["3D-Total"] = diff_xyz_df.pow(2).sum(axis=1, min_count=3).pow(0.5)
+    diff_sp3_df["Clock"] = diff_clk_df
+    diff_sp3_df["|Clock|"] = diff_clk_df.abs()
+
+    # Change the epoch indices from J2000 seconds to more readable python datetimes
+    # and rename the indices properly
+    format_index(diff_sp3_df)
+
+    return diff_sp3_df
+
+
+def clk_difference(
+    base_clk_file: _Path,
+    test_clk_file: _Path,
+    norm_types: list[str],
+) -> _pd.DataFrame:
+    """
+    Compare two CLK files to calculate clock differences with common mode removed (if specified)
+    based on the chosen normalisations.
+
+    :param _Path base_clk_file: Path of the baseline CLK file
+    :param _Path test_clk_file: Path of the test CLK file
+    :param norm_types list[str]: Normalizations to apply. Available options include 'epoch', 'daily', 'sv',
+            any satellite PRN, or any combination of them, defaults to None
+    :return _pd.DataFrame: The Pandas DataFrame containing clock differences
+    """
+    base_clk_df = _gn_io.clk.read_clk(base_clk_file)
+    test_clk_df = _gn_io.clk.read_clk(test_clk_file)
+
+    diff_clk_df = compare_clk(test_clk_df, base_clk_df, norm_types=norm_types)
+
+    # Stack diff_clk_df to keep the format consistent with other dataframes (compare_clk() returns unstacked dataframe)
```

Comment:
generally speaking, comment in the code should be avoid (unless really necessary to explain what's happening.